cuda_bindings/docs/source/overview.md

```diff
@@ -291,22 +290,286 @@ nsys profile -s none -t cuda --stats=true <executable>
 
-Screenshot of Nsight Compute CLI output of CUDA Python example.
+Screenshot of Nsight Compute CLI output of `cuda.bindings` example.
+```
+
+## Preparing kernel arguments
+
+The `cuLaunchKernel` API bindings retain low-level CUDA argument preparation requirements:
+
+* Each kernel argument is a `void*` (i.e. pointer to the argument)
+* `kernelParams` is a `void**` (i.e. pointer to a list of kernel arguments)
+* `kernelParams` arguments are in contiguous memory
+
+These requirements can be met with two different approaches, using either NumPy or ctypes.
+
+### Using NumPy
+
+NumPy [Array objects](https://numpy.org/doc/stable/reference/arrays.html) can be used to fulfill each of these conditions directly.
+
+Let's use the following kernel definition as an example:
+```python
+kernel_string = """\
+typedef struct {
+    int value;
+} testStruct;
+
+extern "C" __global__
+void testkernel(int i, int *pi,
+                float f, float *pf,
+                testStruct s, testStruct *ps)
+{
+    *pi = i;
+    *pf = f;
+    ps->value = s.value;
+}
+"""
+```
+
+The first step is to create array objects with types corresponding to your kernel arguments. Primative Numpy types have the following corresponding kernel types:
+
+```{list-table} Correspondence between NumPy types and kernel types.
+:header-rows: 1
+
+* - NumPy type
+  - Corresponding kernel types
+  - itemsize (bytes)
+* - bool
+  - bool
+  - 1
+* - int8
+  - char, signed char, int8_t
+  - 1
+* - int16
+  - short, signed short, int16_t
+  - 2
+* - int32
+  - int, signed int, int32_t
+  - 4
+* - int64
+  - long long, signed long long
+  - 8
+* - uint8
+  - unsigned char
+  - 1
+* - uint16
+  - unsigned short
+  - 2
+* - uint32
+  - unsigned int
+  - 4
+* - uint64
+  - unsigned long long, uint64_t
+  - 8
+* - float16
+  - half
+  - 2
+* - float32
+  - float
+  - 4
+* - float64
+  - double
+  - 8
+* - complex64
+  - float2, cuFloatComplex, complex&lt;float&gt;
+  - 8
+* - complex128
+  - double2, cuDoubleComplex, complex&lt;double&gt;
+  - 16
+```
+
+Furthermore, custom NumPy types can be used to support both platform-dependent types and user-defined structures as kernel arguments.
+
+This example uses the following types:
+* `int` is `np.uint32`
+* `float` is `np.float32`
+* `int*`, `float*` and `testStruct*` is `np.intp`
+* `testStruct` is a custom user type `np.dtype([("value", np.int32)], align=True)`
+
+Note how both of the pointers are `np.intp` since the pointers values are always a representation of an address space.
+
+Putting it all together:
+```python
+# Define a custom type
+testStruct = np.dtype([("value", np.int32)], align=True)
+
+# Allocate host memory
+pInt_host = checkCudaErrors(cudart.cudaHostAlloc(np.dtype(np.int32).itemsize, cudart.cudaHostAllocMapped))
+pFloat_host = checkCudaErrors(cudart.cudaHostAlloc(np.dtype(np.float32).itemsize, cudart.cudaHostAllocMapped))
+pStruct_host = checkCudaErrors(cudart.cudaHostAlloc(testStruct.itemsize, cudart.cudaHostAllocMapped))
```

Comment:
I suggest that we either switch to just use device memory (which is the most common situation for kernels taking pointers as arguments), or make it clear that we're allocating pinned/page-locked (host) memory here that are accessible by GPU kernels.